### PR TITLE
Update extension test script and add doc for the extension tests

### PR DESCRIPTION
**Pull request by nwmac**
Simple documentation for the extensions tests.

Updates to the test script to better handle error cases and be more resilient, particularly when running locally in dev.

### CODE REVIEW

File: docusaurus/docs/testing/extensions-test.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+### Imports
+
+Importing Vue components using the `~` syntax should not be used - this will build in the case of the dashboard repository, but not in the case of a standalone app.
+
+## Running Tests locally
+
+You can run the script `./shell/scripts/test-plugins-build.sh` locally in a terminal to run the test suite on a developer machine.
+
+> Note: After running the tests the `package.json` files in `shell` and `pkg/rancher-components` will be modified - ensure you reset these and do not check in changes.
```
Comment (cnotv) [inline docusaurus/docs/testing/extensions-test.md:32]: Can we extend the ignore list for these generated files?

```
login.json
pkg/rancher-components/package.json.bak
```

File: docusaurus/docs/testing/extensions-test.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
+# Extensions Tests
+
+The Extensions framework is tested by the GitHub action `check-plugins.yaml`. This runs the script `./shell/scripts/test-plugins-build.sh`
+which runs the actual tests.
+
+The tests check the following:
+
+- Ensures that Verdaccio is installed and running locally (Verdaccio provides a local NPM registry that the tests can use to publish packages without having to publish to npm)
```
Comment (cnotv) [inline docusaurus/docs/testing/extensions-test.md:8]: During the test it requires me credentials, should it be ficticious or which one should I use?
Is it possible (for sure is in CI) to have env vars for this?

When I try with random values I get this errror
```
npm ERR! code E401
npm ERR! Unable to authenticate, your authentication token seems to be invalid.
npm ERR! To correct this please trying logging in again with:
npm ERR!     npm login

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/gleo/.npm/_logs/2023-04-18T08_40_15_742Z-debug.log
error Command failed with exit code 1.
```

Is it meant to be configured in somehow?

File: docusaurus/docs/testing/extensions-test.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+- Verifies the build of the following, by building and then publishing them to the local registry
+  - Shell
+  - Yarn Creator packages
+  - Rancher components package
+- Verifies that a new app can be created with the `@rancher/app` creator and built (test case of a new extension in a new repository)
+- Verifies that an extension can be added with the `@rancher/pkg` creator to the new app and that the dashboard build passes with the new extension present
+- Verifies that the extension can be build standalone
+- Verifies that an extension can be added to the existing dashboard and that the dashboard build passes with the new extension present
```
Comment (cnotv) [inline docusaurus/docs/testing/extensions-test.md:16]: Is there a way to run tasks separately? To test specific case it requires too much time to build everything and run the bundler as well.

File: docusaurus/docs/testing/extensions-test.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+The tests check the following:
+
+- Ensures that Verdaccio is installed and running locally (Verdaccio provides a local NPM registry that the tests can use to publish packages without having to publish to npm)
+- Verifies the build of the following, by building and then publishing them to the local registry
+  - Shell
+  - Yarn Creator packages
+  - Rancher components package
+- Verifies that a new app can be created with the `@rancher/app` creator and built (test case of a new extension in a new repository)
+- Verifies that an extension can be added with the `@rancher/pkg` creator to the new app and that the dashboard build passes with the new extension present
```
Comment (cnotv) [inline docusaurus/docs/testing/extensions-test.md:14]: While creating the pkg it seems like there's some kind of configuration although I see nothing obvious:

```
  Output directory: /home/runner/work/dashboard/dashboard/dist-pkg/test-pkg-0.1.0
~/work/dashboard/dashboard/pkg/test-pkg ~/work/dashboard/dashboard
```

It [generates conflicts](https://github.com/rancher/dashboard/actions/runs/4723080482/jobs/8378536519?pr=8654#step:4:830) on version changes which are probably related to some libraries defined in a `package.json` somewhere, I guess?

**Review by cnotv — CHANGES_REQUESTED** [review]
Added all the parts which I'm facing issues and may probably be helpful to other people as well.

File: docusaurus/docs/testing/extensions-test.md (hunk 4 above)
Comment (cnotv) [reply docusaurus/docs/testing/extensions-test.md:14]: Ok I just found out on my own, the paths are:
- `shell/creators/app`
- `shell/creators/pkg`
- `shell/creators/update`

Could we please add this to the documentation, especially if we have to make changes or upgrades?

File: docusaurus/docs/testing/extensions-test.md (hunk 2 above)
Comment (cnotv) [reply docusaurus/docs/testing/extensions-test.md:8]: I've seen that we have some credentials in the code, which for some reasons are not passed to the command.
When I use them 2 of the 3 login pass, but then it fails on `cd pkg/rancher-components && npm publish` which should be bypassed by Verdaccio:
Screenshot 2023-04-19 at 17 13 38
Screenshot 2023-04-19 at 17 13 47
Screenshot 2023-04-19 at 17 24 43

Comment (cnotv) [reply docusaurus/docs/testing/extensions-test.md:8]: Here's another screenshot where the packages are loaded correctly and I can login as admin (still have to do it manually in CLI) but the final step fails.
I think we can probably use this screenshot as well to display what should happen.

Screenshot 2023-04-20 at 09 50 34

File: docusaurus/docs/testing/extensions-test.md (hunk 1 above)
Comment (cnotv) [reply docusaurus/docs/testing/extensions-test.md:32]: It seems like this is the part of code which we should correct, as it checks the file but not the content:
```
 ~/.config/verdaccio/htpasswd
```

Comment (cnotv) [reply docusaurus/docs/testing/extensions-test.md:32]: If I try to remove the file and run again the command I get
```
[2/4] Logging in...
info npm username: admin
info npm email: admin
question npm password: 
error An unexpected error occurred: "http://localhost:4873/-/user/org.couchdb.user:admin: username is already registered".
```

Comment (cnotv) [reply docusaurus/docs/testing/extensions-test.md:32]: After removing manually the folders for the configuration and libraries the database still seems keeping the users and also the CLI is now broken 💀💀💀💀💀💀💀💀 

Screenshot 2023-04-20 at 11 14 32

Comment (cnotv) [reply docusaurus/docs/testing/extensions-test.md:32]: I was able to make it work after reinstalling it multiple times 🤔

Comment (cnotv) [reply docusaurus/docs/testing/extensions-test.md:32]: If you have to type in the user when running the script locally, you have to logout with `yarn` pointing to the registry and not `npm` as in the script.
E.g. `yarn logout --registry=http://localhost:4873`

Comment (cnotv) [reply docusaurus/docs/testing/extensions-test.md:32]: Although the storage is located in `~/.local/share/verdaccio/storage/.verdaccio-db.json`, it does not seems removing the added `admin` user when changed, so the login always prints out 
```
{
  "error": "username is already registered"
}
```

Comment (cnotv) [reply docusaurus/docs/testing/extensions-test.md:32]: In order to update configurations, the service must be stopped with `kill` command, then restarted the script. This is the only way to also reset the database.

If you want to change the path you can either use an env var (`VERDACCIO_STORAGE_PATH `) or change the `storage:` configuration inside the `config.yaml`

Comment (cnotv) [reply docusaurus/docs/testing/extensions-test.md:32]: Still get 
```
npm ERR! code E401
npm ERR! Unable to authenticate, your authentication token seems to be invalid.
```

File: docusaurus/docs/testing/extensions-test.md (hunk 2 above)
Comment (cnotv) [reply docusaurus/docs/testing/extensions-test.md:8]: To publish a second time you need to change the version, so it may be worthy to mention it somewhere.

Comment (cnotv) [reply docusaurus/docs/testing/extensions-test.md:8]: Changing `npm` command to `yarn` and add the registry seems solve the last step:

```
    "publish:lib": "cd pkg/rancher-components && yarn publish"
```

@nwmac how is it possible that this is different between local and CI?

File: shell/scripts/extension-test/test-plugins-parse-login [hunk 5]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env node
+
+const readline = require('readline');
+
+const rl = readline.createInterface({
```
Comment (cnotv) [inline shell/scripts/extension-test/test-plugins-parse-login:5]: Could you please add a comment about which stream we are interacting as there's no login command around?

File: shell/scripts/test-plugins-build.sh [hunk 6]
```diff
@@ -2,22 +2,132 @@
...
+    fi
+  fi
+
+  if [ ${KEEP_PKG} == false ]; then
+    rm -rf ./pkg/test-pkg
+  fi
+
+  # Restore shell version
+  if [ -n "${SHELL_VERSION}" ]; then
+    sed -i.bak -e "s/\"version\": \"7.7.7\",/\"version\": \"${SHELL_VERSION}\",/g" ${SHELL_DIR}/package.json
```
Comment (cnotv) [inline shell/scripts/test-plugins-build.sh:106]: In case of multiple publications the version value should be increased or Verdaccio throws an error. Although we could clean it up and start again, that's quite time consuming.

File: docusaurus/docs/testing/extensions-test.md [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+- Verifies that the extension can be build standalone
+- Verifies that an extension can be added to the existing dashboard and that the dashboard build passes with the new extension present
+
+## Running Tests locally
+
+You can run the script `./shell/scripts/test-plugins-build.sh` locally in a terminal to run the test suite on a developer machine.
```
Comment (cnotv) [inline docusaurus/docs/testing/extensions-test.md:20]: When everything finally succeed, it creates automatically a commit for the tag:

Screenshot 2023-04-21 at 10 52 44

**Review by cnotv — CHANGES_REQUESTED** [review]
Could we address anyway all the extra information required from all the points?

File: docusaurus/docs/testing/extensions-test.md [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+Run the script with `-h` to see the list of options available.
+
+## Possible causes of test failures
+
+### Dependencies
+
+For the standalone case of a new app using the `@rancher/shell` package, the dependency versions can end up being different to that of the dashboard codebase. This is typically due to dependencies being top-level dependencies in the case of the dashboard codebase, but child dependencies in the case of the new app. In the later case, yarn can hoist dependency versions differently.
+
+To debug this situation, run the test script with `-i -k` - this will only run the standalone tests (`-i` ignores the in-tree tests) and will keep the generate application after the test run (`-k`).
```
Comment (cnotv) [inline docusaurus/docs/testing/extensions-test.md:30]: I get this when I try to run the command `./shell/scripts/test-plugins-build.sh -i -k`

```
Extensions framework build tests
In-tree tests will not be run
Folder for test appliaction will not be deleted after test run
Verdaccio is already running
You should ensure a user is set up and logged in for publishing
npm ERR! code ENEEDAUTH
npm ERR! need auth This command requires you to be logged in.
npm ERR! need auth You need to authorize this machine using `npm adduser`

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/gleo/.npm/_logs/2023-04-25T07_34_36_251Z-debug.log
```

As we use a private registry, this error should not happen and the login should be done by the test.

Using `npm login --registry=http://localhost:4873/` solves this error report.

File: shell/scripts/test-plugins-build.sh [hunk 9]
```diff
@@ -2,22 +2,132 @@
...
+
+usage() {
+  echo "Usage: $0 [<options>] [plugins]"
+  echo " options:"
+  echo "  -k           Keep the folder for the test application after run"
+  echo "  -e           Keep the test extension after run (in-tree tests)"
+  echo "  -i           Skip in-tree tests"
+  echo "  -s           Skip standalone tests"
+  echo "  -x           Skip test setup"
+  echo "  -u <name>    Verdaccio user name for publishing (default: admin)"
```
Comment (cnotv) [inline shell/scripts/test-plugins-build.sh:36]: I am not able to login with this flag:

```
➜  dashboard git:(extension-test-doc) ./shell/scripts/test-plugins-build.sh -u admin -p admin          
Extensions framework build tests
./shell/scripts/test-plugins-build.sh: illegal option -- u
Usage: ./shell/scripts/test-plugins-build.sh [<options>] [plugins]
 options:
  -k           Keep the folder for the test application after run
  -e           Keep the test extension after run (in-tree tests)
  -i           Skip in-tree tests
  -s           Skip standalone tests
  -x           Skip test setup
  -u <name>    Verdaccio user name for publishing (default: admin)
  -p <name>    Verdaccio password for publishing
```

This comment has been added just because the help (cool thing!) is mentioning it but it did not work for me

File: shell/scripts/test-plugins-build.sh [hunk 10]
```diff
@@ -2,22 +2,132 @@
...
+REG_PWD=admin
+
+usage() {
+  echo "Usage: $0 [<options>] [plugins]"
+  echo " options:"
+  echo "  -k           Keep the folder for the test application after run"
+  echo "  -e           Keep the test extension after run (in-tree tests)"
+  echo "  -i           Skip in-tree tests"
+  echo "  -s           Skip standalone tests"
+  echo "  -x           Skip test setup"
```
Comment (cnotv) [inline shell/scripts/test-plugins-build.sh:35]: I get error when using this flag:

```
./shell/scripts/test-plugins-build.sh: illegal option -- x
```

File: shell/scripts/test-plugins-build.sh (hunk 6 above)
Comment (cnotv) [reply shell/scripts/test-plugins-build.sh:106]: We could alternatively add a command after the login where we remove the libraries, as it's quite fast and does not return error if already removed.

The libraries cannot be added in list, unfortunately.

```
npm unpublish --force @rancher/shell --registry=http://localhost:4873/ 
```

**Review by cnotv — CHANGES_REQUESTED** [review]
Except the comments I've been adding, the script seems working fine to me.

File: shell/scripts/test-plugins-build.sh (hunk 6 above)
Comment (nwmac) [reply shell/scripts/test-plugins-build.sh:106]: It should already clean Verdaccio of old packages - its does this directly by deleting files from Verdaccio's local storage.

For npm unpublish, would need to check package is present and version number

Comment (cnotv) [reply shell/scripts/test-plugins-build.sh:106]: If there's no need to add any flag of any type, it seems that the packages are not removed. I had this issue also with the previous version of the script, same message of above.

**Comment by cnotv** [comment]
Could we eventually split the workflow in more steps? It's really hard to follow the whole execution in GH with one Job only.

**Comment by nwmac** [comment]
@cnotv Yes, we could do that, good idea

File: shell/scripts/test-plugins-build.sh (hunk 6 above)
Comment (nwmac) [reply shell/scripts/test-plugins-build.sh:106]: Updated to use npm unpublish now

**Comment by nwmac** [comment]
out of date
